compiler/vm/vmrunner.nim

```diff
@@ -0,0 +1,274 @@
+## This module implements a standalone runner for the VM. The runner is
+## currently only meant for use by the compiler and thus performs only
+## little to no input validation.
+
+import
+  std/[
+    os,
+    strutils
+  ],
+
+  compiler/ast/[
+    ast_types,
+    ast
+  ],
+  compiler/front/[
+    cli_reporter,
+    msgs,
+    options
+  ],
+  compiler/vm/[
+    packed_env,
+    vm,
+    vmdef,
+    vmmemory,
+    vmobjects,
+    vmops,
+    vmtypes
+  ],
+  compiler/ic/[
+    bitabs
+  ],
+  compiler/utils/[
+    pathutils,
+    bitsets
+  ],
+  experimental/[
+    results
+  ]
+
+
+proc loadDiscrConst(s: PackedEnv, constIdx: int, dst: LocHandle,
+                    objTyp: PVmType, fieldPos: FieldPosition): int =
+  let
+    n =            s.nodes[constIdx]
+    packed =       s.numbers[n.pos.LitId]
+    (owner, idx) = getFieldAndOwner(objTyp, fieldPos) # `owner` might be != `objTyp`
+
+  # During const packing, `packDiscr` is used with `numBits` = 32
+  let (v, i) = unpackDiscr(packed, numBits = 32)
+  writeDiscrField(dst, owner, idx, v, i)
+
+  result = 1
+
+proc loadConst(s: PackedEnv, idx: int, dst: LocHandle,
+               mem: var VmMemoryManager): int =
+  ## Loads the constant data represented by the ``PackedDataNode`` sub-tree at
+  ## `idx` into the given VM memory location. Returns the total number of
+  ## nodes that were part of the processed sub-tree
+  let n = s.nodes[idx]
+  case dst.typ.kind
+  of akInt:
+    # this works for both signed and unsigned integers
+    dst.writeUInt(s.getIntVal(n))
+  of akFloat:
+    dst.writeFloat(s.getFloatVal(n))
+  of akPtr:
+    assert n.kind == pdkPtr
+    case n.pos
+    of 0: discard "nil ptr"
+    else: unreachable()
+  of akRef:
+    assert n.kind == pdkPtr
+    case n.pos
+    of 0: discard "nil ref"
+    else: unreachable()
+  of akSeq, akArray:
+    assert n.kind == pdkArray
+    let L = n.pos.int
+
+    if dst.typ.kind == akSeq:
+      deref(dst).seqVal.newVmSeq(dst.typ, L, mem)
+
+    var i = 0
+    while i < L:
+      result += loadConst(s, idx+1+result, getItemHandle(dst, i), mem)
+      inc i
+
+  of akString:
+    assert n.kind == pdkString
+    # TODO: `VmString` needs to be able to store string literals...
```

Comment:
Just confirming are you using `TODO` to indicate handle before the PR is done -- just going by history?

Reply:
I'm mostly using `TODO` to indicate 'handle before PR is done', but sometimes also to document a concrete step that's planned in a refactoring/change split across multiple PRs.